File: SICP/Lecture_2/Code/Python/compound_data/rational_number/rational_number_with_pair.py
from compound_data.ordered_pair.pair import *

def make_rational_number(_numer, _denom):
    g = gcd(_numer, _denom)
    return make_pair(_numer / g, _denom / g)

def numer(rational_number):
    return head(rational_number)

def denom(rational_number):
    return tail(rational_number)

def add(rational_number1, rational_number2):
    return make_rational_number(
        numer(rational_number1) * denom(rational_number2) + denom(rational_number1) * numer(rational_number2),
        denom(rational_number1) * denom(rational_number2))

def mul(rational_number1, rational_number2):
    return make_rational_number(
        numer(rational_number1) * numer(rational_number2),
        denom(rational_number1) * denom(rational_number2))

def gcd(x, y):
    if y == 0:
        return x

    return gcd(y, x % y)

File: SICP/Lecture_2/Code/Java/CompoundData/src/fayelab/sicp/compounddata/rationalnumber/RationalNumberWithPair.java
package fayelab.sicp.compounddata.rationalnumber;

import static fayelab.sicp.compounddata.pair.Pair.*;

import java.util.function.Function;

public class RationalNumberWithPair
{
    public static Function<Integer, Integer> makeRationalNumber(int numer, int denom)
    {
        int gcd = gcd(numer, denom);
        return makePair(numer / gcd, denom / gcd);
    }
    
    public static int numer(Function<Integer, Integer> rationalNumber)
    {
        return head(rationalNumber);
    }

    public static int denom(Function<Integer, Integer> rationalNumber)
    {
        return tail(rationalNumber);
    }
    
    public static Function<Integer, Integer> add(Function<Integer, Integer> rationalNumber1, Function<Integer, Integer> rationalNumber2)
    {
        return makeRationalNumber(numer(rationalNumber1) * denom(rationalNumber2) + denom(rationalNumber1) * numer(rationalNumber2),
                                  denom(rationalNumber1) * denom(rationalNumber2));
    }
    
    public static Function<Integer, Integer> mul(Function<Integer, Integer> rationalNumber1, Function<Integer, Integer> rationalNumber2)
    {
        return makeRationalNumber(numer(rationalNumber1) * numer(rationalNumber2),
                                  denom(rationalNumber1) * denom(rationalNumber2));
    }
    
    public static int gcd(int x, int y)
    {
        if(y == 0)
        {
            return x;
        }
        
        return gcd(y, x % y);
    }
}
